examples/keras_recipes/distillation_recipe.py

```diff
@@ -0,0 +1,518 @@
+
+    dataset = dataset.prefetch(AUTO)
+    return dataset
+
+
+"""
+Note that for brevity, we used mild crops for the training set but in practice
+"Inception-style" preprocessing should be applied. You can refer to
+[this script](https://github.com/sayakpaul/FunMatch-Distillation/blob/main/crop_resize.py)
+for a closer implementation. Also, _**the ground-truth labels are not used during
+training the student.**_
+"""
+
+train_ds = prepare_dataset(train_ds, True)
+validation_ds = prepare_dataset(validation_ds, False)
+test_ds = prepare_dataset(test_ds, False)
+
+"""
+## Visualization
+"""
+
+sample_images, _ = next(iter(train_ds))
+plt.figure(figsize=(10, 10))
+for n in range(25):
+    ax = plt.subplot(5, 5, n + 1)
+    plt.imshow(sample_images[n].numpy())
+    plt.axis("off")
+plt.show()
+
+"""
+## Student model
+
+For the purpose of this example, we will use the standard ResNet50V2 
+([He et al.](https://arxiv.org/abs/1603.05027)).
+"""
+
+
+def get_resnetv2():
+    resnet_v2 = keras.applications.ResNet50V2(
+        weights=None,
+        input_shape=(RESIZE, RESIZE, 3),
+        classes=102,
+        classifier_activation="linear",
+    )
+    return resnet_v2
+
+
+get_resnetv2().count_params()
+
+"""
+Compared to the teacher model, this model has 358 Million less parameters. 
+"""
+
+"""
+## Distillation utility
+
+We will reuse some code from
+[this example](https://keras.io/examples/vision/knowledge_distillation/) on knowledge
+distillation.
+"""
+
+
+class Distiller(tf.keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.student = student
+        self.teacher = teacher
+
+    def compile(
+        self, optimizer, metrics, distillation_loss_fn, temperature=TEMPERATURE,
+    ):
+        super(Distiller, self).compile(optimizer=optimizer, metrics=metrics)
+        self.distillation_loss_fn = distillation_loss_fn
+        self.temperature = temperature
+
+    def train_step(self, data):
+        # Unpack data
+        x, _ = data
+
+        # Forward pass of teacher
+        teacher_predictions = self.teacher(x, training=False)
+
+        with tf.GradientTape() as tape:
+            # Forward pass of student
+            student_predictions = self.student(x, training=True)
+
+            # Compute loss
+            distillation_loss = self.distillation_loss_fn(
+                tf.nn.softmax(teacher_predictions / self.temperature, axis=1),
+                tf.nn.softmax(student_predictions / self.temperature, axis=1),
+            )
+
+        # Compute gradients
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(distillation_loss, trainable_vars)
+
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Report progress
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"distillation_loss": distillation_loss})
```

Comment:
You should add a `Mean` metric to track this quantity, listed in `self.metrics`, otherwise you're only displaying the value from the last batch, not the running average. You can have `self.metrics` return the super value, plus your custom metric (basically mixing the framework-managed metrics and your own).

This makes me think that it could be useful to add automated tracking of metrics on `Model`, as a future Keras feature